#!/bin/bash

# You can control the resources and scheduling with '#SBATCH' settings
# (see 'man sbatch' for more information on setting these parameters)

# The default partition is the 'general' partition
#SBATCH --partition=general

# The default Quality of Service is the 'short' QoS (maximum run time: 4 hours)
#SBATCH --qos=medium

# The default run (wall-clock) time is 1 minute
#SBATCH --time=24:00:00

# The default number of parallel tasks per job is 1
#SBATCH --ntasks=1

# The default number of CPUs per task is 1 (note: CPUs are always allocated to jobs per 2)
# Request 1 CPU per active thread of your program (assume 1 unless you specifically set this)
#SBATCH --cpus-per-task=1

# The default memory per node is 1024 megabytes (1GB) (for multiple tasks, specify --mem-per-cpu instead)
#SBATCH --mem=64GB

# Set mail type to 'END' to receive a mail when the job finishes
# Do not enable mails when submitting large numbers (>20) of jobs at once
#SBATCH --mail-type=END

# fasta.gz file
reference_assembly="[path]/reference-genomes/human/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna.gz"
# fastq.gz file
reads="[path]/real-reads/human/pacbio/hifi/SRR11292120/split.1"
# name of the model (string)
model_name="pacbio_hifi_human2022_error_model"
# paf.gz file generated by minimap2
alignments_file="[path]/alignments/pacbio_hifi_alignments.paf.gz"

srun badread error_model --reference $reference_assembly --reads $reads --alignment $alignments_file > $model_name
